### PR TITLE
Chebyshev or Sigmoid for tanh

### PR DESCRIPTION
**Pull request by gmuraru**
Added the reference for the paper where the polynomial approximations are discussed.
[Here](http://tygert.com/realcrypt.pdf) is the paper (it is also added as a comment)

They talk about more functions and how to approximate those - maybe it is worth to have approximations for more functions?

Added tests and the possibility to choose between Chebyshev or Sigmoid for computing the hyperbolic tangent.

### CODE REVIEW

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 1]
```diff
@@ -243,12 +243,14 @@ def reconstruct(self):
 
         return sy.MultiPointerTensor(children=pointers)
 
-    def zero(self):
+    def zero(self, shape=None):
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py]: Needed this in case I have a matrix multiplication - in that case the dimension for the zero shares might be different.

File: test/torch/tensors/test_precision.py [hunk 2]
```diff
@@ -467,6 +466,34 @@ def test_torch_sigmoid_approx(workers):
...
 
+def test_torch_tanh_approx(workers):
+    """
+    Test the approximate tanh with different tolerance depending on
+    the precision_fractional considered
+    """
+    alice, bob, james = workers["alice"], workers["bob"], workers["james"]
+
+    fix_prec_tolerance_by_method = {
+        "chebyshev": {3: 3 / 100, 4: 3 / 100, 5: 3 / 100},
```
Comment (gmuraru) [inline test/torch/tensors/test_precision.py]: Those might need more tweaking - currently, it works, but we might want a lower bound

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 3]
```diff
@@ -347,6 +349,8 @@ def add(self, shares: dict, other):
                 crypto_provider=self.crypto_provider,
                 **no_wrap,
             ).child
+        elif isinstance(other, AdditiveSharingTensor):
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py]: This is not supposed to happen: the decorator makes sure that the AST other is replaced by his .child

So if you still have an AST it means that you've operated something like [syfttype]>AST>... with the self AST

What is the usecase for this?

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 4]
```diff
@@ -482,11 +486,18 @@ def _public_mul(self, shares, other, equation):
...
+
+                it = iter(shares.items())
+                worker, share = next(it)
+                first_cmd = cmd(share, other)
+                zero_shares = self.zero(first_cmd.shape).child
+                res[worker] = (first_cmd + zero_shares[worker]) % self.field
+
+                for worker, share in it:
+                    cmd_res = cmd(share, other)
+                    res[worker] = (cmd(share, other) + zero_shares[worker]) % self.field
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py]: Can you put all in the same loop  (and add a flag to notice when it's the first iteration) ?

File: syft/frameworks/torch/tensors/interpreters/precision.py [hunk 5]
```diff
@@ -413,7 +413,7 @@ def matmul(self, *args, **kwargs):
             ), "In matmul, all args should have the same precision_fractional"
 
         if isinstance(self.child, AdditiveSharingTensor) and isinstance(other.child, torch.Tensor):
-            # If we try to matmul a FPT>(wrap)>AST with a FPT>torch.tensor,
+            # If we try to matmul a FPT>(wrap)>AST with a FPT>(wrap)>torch.tensor,
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/precision.py]: Nope, `FPT>(wrap)>torch.tensor` is forbidden!

But actually `FPT>(wrap)>AST` is not correct anymore, now we do `FPT>AST`  directly

File: syft/frameworks/torch/tensors/interpreters/precision.py [hunk 6]
```diff
@@ -422,7 +422,7 @@ def matmul(self, *args, **kwargs):
         elif isinstance(other.child, AdditiveSharingTensor) and isinstance(
             self.child, torch.Tensor
         ):
-            # If we try to matmul a FPT>torch.tensor with a FPT>(wrap)>AST,
+            # If we try to matmul a FPT>(wrap)>torch.tensor with a FPT>(wrap)>AST,
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/precision.py]: same here

File: test/torch/tensors/test_precision.py (hunk 2 above)
Comment (LaRiffle) [reply test/torch/tensors/test_precision.py]: That's ok I guess!

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py (hunk 4 above)
Comment (gmuraru) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py]: Sure

File: syft/frameworks/torch/tensors/interpreters/precision.py [hunk 7]
```diff
@@ -422,7 +422,7 @@ def matmul(self, *args, **kwargs):
         elif isinstance(other.child, AdditiveSharingTensor) and isinstance(
             self.child, torch.Tensor
         ):
-            # If we try to matmul a FPT>torch.tensor with a FPT>(wrap)>AST,
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/precision.py]: @LaRiffle removed the wrap

Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: Great!
